subsys/usb/usb_c/usbc_pe_src_states.c

```diff
@@ -0,0 +1,702 @@
+ */
+void pe_src_capability_response_entry(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+
+	/*
+	 * On entry to the PE_SRC_Capability_Response state the Policy Engine
+	 * Shall request the Protocol Layer to send one of the following:
+	 */
+
+	/*
+	 * 1: Reject Message – if the request cannot be met or the present
+	 *    Contract is Invalid.
+	 */
+	if (pe->snk_request_reply == SNK_REQUEST_REJECT) {
+		pe_send_ctrl_msg(dev, PD_PACKET_SOP, PD_CTRL_REJECT);
+	}
+	/*
+	 * 2: Wait Message – if the request could be met later from the Power
+	 *    Reserve. A Wait Message Shall Not be sent if the present Contract
+	 *    is Invalid.
+	 */
+	else {
+		pe_send_ctrl_msg(dev, PD_PACKET_SOP, PD_CTRL_WAIT);
+	}
+}
+
+void pe_src_capability_response_run(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+
+	/*
+	 * The Policy Engine Shall transition to the PE_SRC_Ready state when:
+	 *	1: There is an Explicit Contract AND
+	 *	2: A Reject Message has been sent and the present Contract
+	 *	   is still Valid OR
+	 *	3: A Wait Message has been sent.
+	 */
+	if (atomic_test_bit(pe->flags, PE_FLAGS_EXPLICIT_CONTRACT) &&
+			((pe->snk_request_reply == SNK_REQUEST_REJECT &&
+			policy_check_present_contract(dev, pe->present_contract)) ||
+			(pe->snk_request_reply == SNK_REQUEST_WAIT))) {
+		pe_set_state(dev, PE_SRC_READY);
+	}
+	/*
+	 * The Policy Engine Shall transition to the PE_SRC_Hard_Reset state
+	 * when:
+	 *	1: There is an Explicit Contract and
+	 *	2: The Reject Message has been sent and the present Contract
+	 *	   is Invalid
+	 */
+	else if (atomic_test_bit(pe->flags, PE_FLAGS_EXPLICIT_CONTRACT) &&
+			policy_check_present_contract(dev, pe->present_contract) == false) {
+		pe_set_state(dev, PE_SRC_HARD_RESET);
+	}
+	/*
+	 * The Policy Engine Shall transition to the PE_SRC_Wait_New_Capabilities
+	 * state when:
+	 *	1: There is no Explicit Contract and
+	 *	2: A Reject Message has been sent or
+	 *	3: A Wait Message has been sent.
+	 */
+	else {
+		/* 8.3.3.2.13 PE_SRC_Wait_New_Capabilities embedded here */
+
+		/*
+		 * In the PE_SRC_Wait_New_Capabilities State the Device Policy Manager
+		 * Should either decide to send no further Source Capabilities or
+		 * Should send a different set of Source Capabilities. Continuing
+		 * to send the same set of Source Capabilities could result in a live
+		 * lock situation.
+		 */
+
+		/* Notify DPM to send a different set of Source Capabilities */
+		if (policy_change_src_caps(dev)) {
+			/* DPM will send different set of Source Capabilities */
+			pe_set_state(dev, PE_SRC_SEND_CAPABILITIES);
+		} else {
+			/*
+			 * DPM can not send a different set of Source
+			 * Capabilities, so disable port.
+			 */
+			pe_set_state(dev, PE_SUSPEND);
+		}
+	}
+}
+
+/**
+ * @brief 8.3.3.2.9 PE_SRC_Hard_Reset State
+ */
+void pe_src_hard_reset_entry(void *obj)
```

Comment:
It seems like PE_SRC_Hard_Reset_Received and PE_SRC_Hard_Reset could share most of their code.

Reply:
I moved common code to a parent state

Reply:
It looks like you made `PE_SRC_HARD_RESET_RECEIVED` its own parent with no implementation. Should it have `PE_SRC_HARD_RESET_PARENT` as its parent?

Reply:
Good catch.